### PR TITLE
Source location in expressions

### PR DESCRIPTION
**Pull request by gzanitti**
This PR solves issue #1293.
I am using the SourceRef defined in the AST, but it could be changed. It might even be more convenient to use Option<SourceRef> so that we don't have to use SourceRef::unknown() in some cases. Let me know what you think.

All comments are welcome :)

### CODE REVIEW

File: executor/src/constant_evaluator/mod.rs [hunk 1]
```diff
@@ -69,10 +70,19 @@ fn generate_values<T: FieldElement>(
...
-                });
+                index_expr = Expression::IndexAccess(
+                    IndexAccess {
+                        array: e.clone().into(),
+                        index: Box::new(Expression::Number(
+                            Number {
+                                value: index.into(),
+                                type_: None,
+                            },
+                            SourceRef::unknown(),
```
Comment (gzanitti) [inline executor/src/constant_evaluator/mod.rs:81]: I'm not sure about these two unknowns, should I take the value from e? Or does it not make sense in this context?

File: executor/src/constant_evaluator/mod.rs [hunk 2]
```diff
@@ -69,10 +70,19 @@ fn generate_values<T: FieldElement>(
             };
             let index_expr;
             let e = if let Some(index) = index {
-                index_expr = Expression::IndexAccess(IndexAccess {
-                    array: e.clone().into(),
-                    index: Box::new(Expression::Number(index.into(), None)),
-                });
+                index_expr = Expression::IndexAccess(
+                    SourceRef::unknown(),
+                    IndexAccess {
```
Comment (chriseth) [inline executor/src/constant_evaluator/mod.rs:75]: I'm wondering if it  makes sense to implement Into from IndexAccess to Expression, where the source ref is set to unknown (and of course for all other structs inside expression). Could that lead to us not setting the source ref even though it is known?

File: ast/src/parsed/mod.rs [hunk 3]
```diff
@@ -332,29 +333,163 @@ impl<Expr> Children<Expr> for SelectedExpressions<Expr> {
...
-    IfExpression(IfExpression<Self>),
-    BlockExpression(Vec<StatementInsideBlock<Self>>, Box<Self>),
+    Reference(SourceRef, Ref),
+    PublicReference(SourceRef, String),
+    Number(SourceRef, Number),
+    String(SourceRef, String),
+    Tuple(SourceRef, Vec<Self>),
+    LambdaExpression(SourceRef, LambdaExpression<Self>),
+    ArrayLiteral(SourceRef, ArrayLiteral<Self>),
+    BinaryOperation(SourceRef, BinaryOperation<Self>),
```
[the remark this answers is not in the document]
Comment (gzanitti) [reply ast/src/parsed/mod.rs:345]: Sure!

File: parser/src/lib.rs [hunk 4]
```diff
@@ -249,20 +258,164 @@ mod test {
 
     fn pil_statement_clear_source_ref(stmt: &mut PilStatement) {
         match stmt {
-            PilStatement::Include(s, _)
```
Comment (chriseth) [inline parser/src/lib.rs:252]: Could this function use Children trait?

File: parser/src/powdr.lalrpop [hunk 5]
```diff
@@ -450,8 +450,8 @@ BoxedExpression: Box<Expression> = {
 }
 
 LambdaExpression: Box<Expression> = {
-    <kind:FunctionKind> "||" <body:BoxedExpression> => Box::new(Expression::LambdaExpression(LambdaExpression{kind, params: vec![], body})),
-    <kind:FunctionKind> "|" <params:ParameterList> "|" <body:BoxedExpression> => Box::new(Expression::LambdaExpression(LambdaExpression{kind, params, body})),
+    <start:@L> <kind:FunctionKind> "||" <body:BoxedExpression> => Box::new(Expression::LambdaExpression(ctx.source_ref(start), LambdaExpression{kind, params: vec![], body})),
```
Comment (chriseth) [inline parser/src/powdr.lalrpop:453]: I think we should also store the end of the expression, but that can be done in another PR.

File: ast/src/parsed/mod.rs [hunk 6]
```diff
@@ -332,29 +333,163 @@ impl<Expr> Children<Expr> for SelectedExpressions<Expr> {
...
+            (Expression::MatchExpression(_, a), Expression::MatchExpression(_, b)) => a == b,
+            (Expression::IfExpression(_, a), Expression::IfExpression(_, b)) => a == b,
+            (Expression::BlockExpression(_, a), Expression::BlockExpression(_, b)) => a == b,
+            _ => false,
+        }
+    }
+}
+
+pub trait SourceInfo {
+    fn get_source(&self) -> &SourceRef;
```
Comment (chriseth) [inline ast/src/parsed/mod.rs:380]: Shouldn't this just be `source(&self)`? Also it's not really the source, it's a source reference, but maybe we can drop that distinction.

File: ast/src/parsed/mod.rs [hunk 7]
```diff
@@ -332,29 +333,163 @@ impl<Expr> Children<Expr> for SelectedExpressions<Expr> {
...
+    fn get_source_mut(&mut self) -> &mut SourceRef;
+    fn set_source(&mut self, source: SourceRef);
```
Comment (chriseth) [inline ast/src/parsed/mod.rs:382]: I think this is not used at all. I think in the rare situation where we need it, we can just use `*x.get_source_mut() = s;` (or `*x.source_mut()  = s;`

File: parser/src/powdr.lalrpop [hunk 8]
```diff
@@ -514,7 +514,7 @@ BinaryAndOp: BinaryOperator = {
 }
 
 BitShift: Box<Expression> = {
-    BitShift BitShiftOp Sum => Box::new(Expression::BinaryOperation(<>)),
+    <start:@L> <left:BitShift> <op:BitShiftOp> <right:Sum> => Box::new(Expression::BinaryOperation(ctx.source_ref(start), BinaryOperation { left, op, right })),
```
Comment (chriseth) [inline parser/src/powdr.lalrpop:517]: I'm wondering if we could simplify this into
```suggestion
    <start:@L> <left:BitShift> <op:BitShiftOp> <right:Sum> => Box::new((BinaryOperation { left, op, right }).into().with_source(ctx.source_ref(start))),
```

if we have Into for conversion between the Expression sub-structs into Expression (and unknown source) and then a generic function on Expression:

```
impl Expression {
  pub fn with_source(&mut self, s: SourceRef) -> Self {
    *(self.source_mut()) = s;
    self
  }
}
```

**Comment by gzanitti** [comment]
Hey @chriseth, ended up separating all the structs in a set of PRs to make them easy to review (#1351 #1352 #1353 #1354 #1355). There is also a PR that adds the SourceRefs on top of them (#1357). I will close this.

**Pull request closed without merging**